sdk/src/main/java/tech/relaycorp/sdk/GatewayClientImpl.kt

```diff
@@ -0,0 +1,211 @@
+package tech.relaycorp.sdk
+
+import android.content.Context
+import android.content.ServiceConnection
+import android.os.Handler
+import android.os.Looper
+import android.os.Messenger
+import android.util.Log
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.channels.BroadcastChannel
+import kotlinx.coroutines.delay
+import kotlinx.coroutines.flow.Flow
+import kotlinx.coroutines.flow.asFlow
+import kotlinx.coroutines.flow.collect
+import kotlinx.coroutines.withContext
+import tech.relaycorp.poweb.PoWebClient
+import tech.relaycorp.relaynet.bindings.pdc.Signer
+import tech.relaycorp.relaynet.bindings.pdc.StreamingMode
+import tech.relaycorp.relaynet.issueParcelDeliveryAuthorization
+import tech.relaycorp.relaynet.messages.Parcel
+import tech.relaycorp.relaynet.messages.control.PrivateNodeRegistrationRequest
+import tech.relaycorp.relaynet.wrappers.x509.Certificate
+import tech.relaycorp.sdk.background.suspendBindService
+import java.security.KeyPair
+import java.time.Duration
+import kotlin.coroutines.resume
+import kotlin.coroutines.resumeWithException
+import kotlin.coroutines.suspendCoroutine
+
+class GatewayClientImpl(
+    private val context: Context
+) {
+
+    // Gateway
+
+    private var syncConnection: ServiceConnection? = null
+
+    suspend fun bind() {
+        withContext(Dispatchers.IO) {
+            if (syncConnection != null) return@withContext // Already connected
+
+            val bindResult = context.suspendBindService(
+                Relaynet.GATEWAY_PACKAGE,
+                Relaynet.GATEWAY_SYNC_COMPONENT
+            )
+            syncConnection = bindResult.first
+            delay(1_000) // Wait for server to start
+        }
+    }
+
+    fun unbind() {
+        syncConnection?.let { context.unbindService(it) }
+        syncConnection = null
+    }
+
+    // First-Party Endpoints
+
+    suspend fun registerEndpoint(keyPair: KeyPair): Pair<Certificate, Certificate> {
```

Comment:
I guess this can be `internal`, since it's only used by `FirstPartyEndpoint.create()`.